design/client-initiated-backchannel-authentication-flow.md

```diff
@@ -0,0 +1,494 @@
+
+### Authentication Flow
+
+Like other flows (Browser Flow, Direct Grant Flow, etc.), the authentication flow for CIBA flow is also needed to execute all the processes that need to be executed after the end-user logged-in like creating User Session and Authenticated Client Session that are needed to generate tokens in Token Req/Res part.
+
+However, authentication and authorization is executed by AuthN/AuthZ by AD part so that it is enough for the authentication flow for CIBA flow to contain only the pass-through authenticator. 
+
+### Event
+
+The follwing events are newly added :
+
+* AUTH_REQ_ID_TO_TOKEN : On token endpoint, keycloak successfully generate tokens in return to "auth_req_id".
+* AUTH_REQ_ID_TO_TOKEN_ERROR : On token endpoint, keycloak fails to generate tokens in return to "auth_req_id".
+
+### Server Metadata
+
+The supported server metadata is as follows :
+
+* "backchannel_token_delivery_modes_supported" : ["poll"]
+* "backchannel_authentication_endpoint" : "{scheme}://{host}[:{port}]/auth/realms/{realm}/protocol/openid-connect/backchannelAuthn"
+* "backchannel_user_code_parameter_supported" : false
+
+### Client Metadata
+
+The supported client metadata is as follows :
+
+* "backchannel_token_delivery_mode" : only accepts "poll"
+
+### Token Request Throttling
+
+According CIBA protocol specification, keycloak determines whether it refuses the token request based on its accompanying "auth_req_id".
+
+There are three options considered :
+
+1. Utilize existing `DefaultBruteForceProtector`
+2. Use Inifinispan cache with lifespan of its entry to share among all nodes in the cluster information on which token request with "auth_req_id" is refused
+3. Use some collection with lifespan of its entry on each node, not sharing among all nodes in the cluster information on which token request with "auth_req_id" is refused
+
+1 is too rich for this token request throttling.
+
+2 can satisfies CIBA protocol specification but has performance concern if vast number of token requests happens. When the token request arrives without waiting an interval, keycloak needs to update when this next token request with "auth_req_id" be allowed to be accepted with some additional wait as penalty. Updating the entry of Infinispan cache causes some traffic among nodes in the cluster to sync the status of this entry.
+
+3 can partially satisfies CIBA protocol specification but no traffic is exchanged among nodes in the cluster due to this token request throttling. However, it seems that keycloak does not prepare such collection so that implementing it or introducing some kind of OSS is required.
+
+In this time, 2 is to be realized.
+
+## Extensibility
+
+In order for users to use CIBA flow on their each use-case, the following SPIs are introduced.
+
+### AuthN/AuthZ User Resolver
+
+This provider converts user information in "login_hint", "login_hint_token" and "id_token_hint" to User Model.
```

Comment:
As for the way of identifying the end user to be authenticated by AD, I'd like to support `login_hint` and `id_token_hint` as I've commented later on.

As for the contents of `login_hint`,  as you pointed out, it is not appropriate to use "public" known value like phone number, user name and email. IMO it is appropriate to use the value known only by the client and keycloak. However, I think such the value is case-by-case so that keycloak itself cannot implement actual resolver. As a compromise, keycloak implement resolvers for username and email but mention on the document that these resolver should only be used for test-use, not be used in service. WDYT?

As for clients that can be allowed to use CIBA flow, we can control which client be allowed to use CIBA flow using [Client  Policies](https://github.com/keycloak/keycloak-community/blob/master/design/client-policies.md) . I'd like to add this description about it on this PR's design document. WDYT?

As for `user_code`, I agree with you. I'd like to add to this PR's design documentation that the resolver for `user_code` is also supported. WDYT?

Reply:
I agree we should advertize the risks when using public known values. The `id_token_hint` seems more natural for me but it would require the client to keep the id_token for a user and I'm not sure if all clients using CIBA would be able to authenticate their users beforehand in order to get the token. I understand that the different paths for identifying users are stricly related on the use-case and how the client can pass the information to the OP?

I also think that the user code is useful so users can provide their user codes during registration or through the account console (or even by setting this from your "management" system through our APIs) so that when resolving the user code we match it based on some user attribute (which can potentially be set on a per-realm or per-client basis).

What about `login_hint_token`? Could we use it to communicate values that are encrypted or signed? It looks like a good option too.

I'm not sure about Client Profiles. It is more related to enforcing policies instead of enabling or disabling features?

Reply:
@pedroigor 
>I'm not sure about Client Profiles. It is more related to enforcing policies instead of enabling or disabling features?

Yes, but this Client Policies' design document in under discussion.

Reply:
IMO, there is 2 types of representation for conveying user to be authenticated from a client (CD) to keycloak :
- public or guessable
  can be acquired from public sources (e.g. name, phone number)
  even if cannot be acquired from public sources, easy to guess (e.g. email address)
- not public and not guessable
  cannot be acquired from public source and hard to guess (e.g. UserModel's ID in keycloak internal)

Considering security, we pay attention to the following points :
- Unsolicited backchannel authentication request by a malicious user (impersonation)
    - public or guessable user information is vulnerable to this attack.
    - Regardless of the way of conveying user to be authenticated, User Code seems to be effective countermeasure.

- Unsolicited backchannel authentication request by a malicious other client to attack other victim client (replay, forge legitimate backchannel authentication request)
    - PPID seems to be effective countermeasure.
    - Encryption seems to be effective countermeasure.
    - Signed JWT seems to be effective countermeasure.
    - Regardless of the way of conveying user to be authenticated, User Code seems to be effective countermeasure.

- Unsolicited backchannel authentication request by a malicious client itself (keep user's user_code and other authentication related information once this client tried CIBA flow with this user, forge legitimate backchannel authentication request)
    - No principle effective countermeasure by keycloak seems not to exist (but might be my misunderstanding). Using binding message and the end user being taking extra precaution on it seems to be only the way to prevent it.

- PII leakage on the wire
    - PPID seems to be effective countermeasure.
    - Encryption seems to be effective countermeasure.

Considering these points, my proposal is as follows :

- `user_code`
  - supported

- `login_hint`
  - supported
  - Resolver : keycloak provides the resolver for the following information
    - User Information
      - username
      - email
  - Encryption : this resolver support encryption for them by using client secret with AES128/192/256bit

- `login_hint_token`
  - supported
  - Resolver : keycloak provides the resolver for the following token
    - Format : JWT
    - Profile : TBD but now considering SET
    - User Information
      - username
      - email
   - Integrity Protection/Message Authentication : JWS
   - Encryption : JWE

- `id_token_hint`
  - supported
  - PPID support for "sub"

WDYT?

Reply:
Also @pedroigor pointed out, there is the case that the user registers multiple ADs. Therefore, the client needs to specify which AD is used for user authentication. CIBA specification seems not to mention it so that we need to specify it as keycloak specific way. If we treat only single AD per user, we does not need consider this issue. Should we support only support single AD or multiple ADs? If supporting multiple ADs, should we support it as a follow-up other PR?

Reply:
@tnorimat +1 to your proposal. Regarding support for multiple ADs, I vote to do it as a follow-up if possible.

In general, it will be good to divide this work into bigger amount of smaller pull requests (due the easier review etc), instead of adding one super-big PR :-) Of course, there can be always exceptions to this rule (For example if there will be additional issues with the migration in case that initial PR and follow-up PR will be done in different Keycloak releases etc).

Reply:
@mposolda I see, I'd like to divide this work into several pull requests and JIRA tickets.

Reply:
@mposolda As for support for multiple ADs, I agree with you. I'd like to realize it as a follow-up task.